### PR TITLE
Call chainerx.grad from chainer.grad

### PR DESCRIPTION
**Pull request by emcastillo**
Addresses PART of #6191 

The added code converts back and forth from the chainer to chainerx interface

ChainerX grad method has a more strict interface where some of the optional arrays are not broadcasted as in chainer.

If output_grads are not of the same shape of the output variable function chainerx will complain while chainer backprop does not.

i.e. for Function.Sum a `chainerx.array(1.0, ...)` as output_grad works while  `chainerx.array([1.0], ...)` does not when compared to chainer grad.

This PR adds support for `input_grads`, `output_grads`, `set_grad` and `retain_grad` options in `chainerx.Grad`

### CODE REVIEW

File: chainer/function_node.py [hunk 1]
```diff
@@ -1009,6 +1009,44 @@ def grad(outputs, inputs, grad_outputs=None, grad_inputs=None, set_grad=False,
                 'len(inputs) = {}, len(grad_inputs) = {}'
                 .format(len(inputs), len(grad_inputs)))
 
+    # Check if all the inputs are chainerx arrays and if so
+    # Relies in chainerx.grad function
+    chx_inputs = sum(map(lambda x: x._has_chainerx_array
```
Comment (niboshi) [inline chainer/function_node.py]: How about this?
`sum([False if x is None else x._has_chainerx_array for x in inputs])`

Comment (niboshi) [inline chainer/function_node.py]: How about `n_chx_inputs` to make it clear that it's a number?

File: chainer/function_node.py [hunk 2]
```diff
@@ -1009,6 +1009,44 @@ def grad(outputs, inputs, grad_outputs=None, grad_inputs=None, set_grad=False,
...
+                         if x is not None else False, inputs))
+    if chx_inputs == len(inputs):
+        # Need to access the arrays to invoke the chainer grad function
+        if grad_inputs is not None:
+            grad_inputs_chx = list(map(lambda x: x._data[0], grad_inputs))
```
Comment (niboshi) [inline chainer/function_node.py]: `[x._data[0] for x in grad_inputs]`

File: chainer/function_node.py [hunk 3]
```diff
@@ -1009,6 +1009,44 @@ def grad(outputs, inputs, grad_outputs=None, grad_inputs=None, set_grad=False,
...
+        else:
+            grad_inputs_chx = []
+        if grad_outputs is not None:
+            grad_outs_chx = list(map(lambda x: x._data[0], grad_outputs))
```
Comment (niboshi) [inline chainer/function_node.py]: ditto.

File: chainer/function_node.py [hunk 4]
```diff
@@ -1009,6 +1009,44 @@ def grad(outputs, inputs, grad_outputs=None, grad_inputs=None, set_grad=False,
...
+        else:
+            grad_outs_chx = []
+        if retain_grad:
+            raise ValueError(
+                'retain_grad is not supported on chainerx.grad interface')
+        if loss_scale is not None:
+            raise ValueError(
+                'loss_scale is not supported on chainerx.grad interface')
+
+        outputs_chx = list(map(lambda x: x._data[0], outputs))
```
Comment (niboshi) [inline chainer/function_node.py]: ditto.

File: chainer/function_node.py [hunk 5]
```diff
@@ -1009,6 +1009,44 @@ def grad(outputs, inputs, grad_outputs=None, grad_inputs=None, set_grad=False,
...
+        inputs_chx = list(map(lambda x: x._data[0], inputs))
```
Comment (niboshi) [inline chainer/function_node.py]: ditto.

**Comment by niboshi** [comment]
@emcastillo Is it ready for review?

**Comment by emcastillo** [comment]
@niboshi I added tests on the chainerx side, there is a lack of tests on the chainer side 
(need to re-think the Grad existing test to work with chainerx) but the rest of the code should be reviewable

**Comment by emcastillo** [comment]
@niboshi chainerx cases added for chainer grad test.
Should be reviewable now

File: chainerx_cc/chainerx/backward.cc [hunk 6]
```diff
@@ -395,8 +397,10 @@ class BackwardImpl {
                 output_array_nodes_.end()) {
                 if (output_array_node != nullptr) {
                     std::shared_ptr<ArrayBody> body = output_array_node->weak_body().lock();
-                    if (body != nullptr && !body->IsGradRequired(backprop_id_)) {
+                    if (body != nullptr && !retain_grad_ && !body->IsGradRequired(backprop_id_)) {
```
Comment (niboshi) [inline chainerx_cc/chainerx/backward.cc]: * `body->IsGradRequired()` can be called twice.
* In the second evaluation, `body` might be `nullptr`. (test condition is insufficient?)

How about below?:
```c++
if (body != nullptr && !body->IsGradRequired(backprop_id_)) {
    if (retain_grad_) {
        internal::ArrayBody::RequireGrad(body, backprop_id_);
    } else {
        body->ClearGrad(backprop_id_);
    }
    
}
```

File: chainerx_cc/chainerx/backward.cc [hunk 7]
```diff
@@ -590,39 +597,67 @@ std::vector<nonstd::optional<Array>> Grad(
...
+        const std::vector<ConstArrayRef>& grad_inputs,
+        const std::vector<ConstArrayRef>& grad_outputs) {
     if (inputs.empty()) {
         return {};
     }
     if (outputs.empty()) {
         return std::vector<nonstd::optional<Array>>(inputs.size(), nonstd::nullopt);
     }
 
+    if (grad_inputs.size() && (inputs.size() != grad_inputs.size())) {
```
Comment (niboshi) [inline chainerx_cc/chainerx/backward.cc]:
```suggestion
    if (!grad_inputs.empty() && inputs.size() != grad_inputs.size()) {
```

* Parentheses are redundant
* Do not evaluate a non-bool value as bool.

File: chainerx_cc/chainerx/backward.cc [hunk 8]
```diff
@@ -590,39 +597,67 @@ std::vector<nonstd::optional<Array>> Grad(
...
+        throw GradientError{"Grad inputs and inputs must have the same size"};
+    }
+    if (grad_outputs.size() && (outputs.size() != grad_outputs.size())) {
```
Comment (niboshi) [inline chainerx_cc/chainerx/backward.cc]: Ditto.

File: chainerx_cc/chainerx/backward.cc [hunk 9]
```diff
@@ -590,39 +597,67 @@ std::vector<nonstd::optional<Array>> Grad(
...
     // Initialize the grad map with newly created gradient arrays of the inputs.
     // The existing gradients of the inputs are thus not modified.
     std::unordered_map<ArrayNode*, internal::GradRef> array_node_grad_map;
-    for (const Array& input : inputs) {
-        const std::shared_ptr<ArrayBody>& array_body = internal::GetArrayBody(input);
+    for (size_t i = 0; i < inputs.size(); i++) {
+        const std::shared_ptr<ArrayBody>& array_body = internal::GetArrayBody(inputs[i]);
         if (const std::shared_ptr<ArrayNode>& input_array_node = array_body->GetArrayNode(actual_backprop_id)) {
-            input_grads.emplace_back(nonstd::optional<Array>{});
+            input_grads.emplace_back(grad_inputs.size() ? nonstd::optional<Array>{grad_inputs[i]} : nonstd::optional<Array>{});
```
Comment (niboshi) [inline chainerx_cc/chainerx/backward.cc]: ditto.

File: chainerx_cc/chainerx/backward.cc [hunk 10]
```diff
@@ -590,39 +597,67 @@ std::vector<nonstd::optional<Array>> Grad(
         const std::vector<ConstArrayRef>& outputs,
         const std::vector<ConstArrayRef>& inputs,
         const nonstd::optional<BackpropId>& backprop_id,
-        DoubleBackpropOption double_backprop) {
+        DoubleBackpropOption double_backprop,
+        bool set_grad,
+        bool retain_grad,
+        const std::vector<ConstArrayRef>& grad_inputs,
```
Comment (niboshi) [inline chainerx_cc/chainerx/backward.cc]: I wonder if we can omit `grad_inputs` argument: https://github.com/chainer/chainer/issues/5978#issuecomment-453090989

**Comment by niboshi** [comment]
This PR fixes #6316

File: chainerx_cc/chainerx/backward.cc (hunk 10 above)
Comment (emcastillo) [reply chainerx_cc/chainerx/backward.cc]: I agree with removing it, I just added to make chainer and chainerx grad interface to be consistent.

Comment (niboshi) [reply chainerx_cc/chainerx/backward.cc]: Please remove it for now. We may add it later if it's really needed.

Comment (emcastillo) [reply chainerx_cc/chainerx/backward.cc]: Ok!

**Comment by emcastillo** [comment]
When removing grad_inputs (a list of variables to initialize the values of the gradients) as a parameter of chainerx.grad.

I tried to modify chainer.grad to just add the values of grad_inputs to the gradients returned by chainerx.grad. However, in some cases chainerx.grad returned arrays do not have the require_gradient flag set, but the grad_inputs values do, resulting on errors when creating a variable by adding the grad_input corresponding offset to the gradient.

```
grad_inputs_chx = [x._data[0] for x in grad_inputs]
[variable.Variable(g+gi, requires_grad=g.is_backprop_required()) 
for g, gi in zip(grads, grad_inputs_chx)]
```
in some cases (double_backprop?) `g` has the `is_backprop_required` flag set
and in some other cases `gi` has it so I wonder if there is some convention to deal with this kind of issue.

**Comment by niboshi** [comment]
I thinks `requires_grad` option can be `g.is_backprop_required() or gi.is_backprop_required()`.
If backprop is required in any of the inputs of `+`, the output should also have backprop required in order to calculate the inputs' gradients.

**Comment by emcastillo** [comment]
I solved it by just using the additions in a temporal array, then the is_back_prop of the addition result has now the correct value and I feel it cleaner :)

File: chainerx_cc/chainerx/backward.cc [hunk 11]
```diff
@@ -621,13 +639,28 @@ std::vector<absl::optional<Array>> Grad(
...
+        const std::shared_ptr<ArrayBody>& array_body = internal::GetArrayBody(outputs[i]);
+        if (const std::shared_ptr<ArrayNode>& array_node = array_body->GetArrayNode(actual_backprop_id)) {
+            output_grads.emplace_back(grad_outputs[i]);
+            array_node_grad_map.emplace(array_node.get(), internal::GradRef{&output_grads.back()});
+        }
+    }
+
+    BackwardImpl{inputs, outputs, actual_backprop_id, double_backprop, std::move(array_node_grad_map), retain_grad}.Run();
 
+    size_t i = 0;
```
Comment (niboshi) [inline chainerx_cc/chainerx/backward.cc]: Why using different index than that of `input_grads`?
(In other words in which case `grad.has_value()` evaluates to false?)

File: chainerx_cc/chainerx/backward.cc [hunk 12]
```diff
@@ -621,13 +639,28 @@ std::vector<absl::optional<Array>> Grad(
...
     // Those grads are returned as nullopt.
     for (absl::optional<Array>& grad : input_grads) {
-        if (grad.has_value() && internal::GetArrayBody(*grad) == nullptr) {
-            grad = absl::nullopt;
+        if (grad.has_value()) {
+            if (internal::GetArrayBody(*grad) == nullptr) {
+                grad = absl::nullopt;
+            }
+            if (set_grad) {
+                inputs[i++].get().SetGrad(grad.value(), backprop_id);
```
Comment (niboshi) [inline chainerx_cc/chainerx/backward.cc]: Let's avoid `i++` embedded in an expression, for readability.

File: tests/chainer_tests/test_function_node.py [hunk 13]
```diff
@@ -920,6 +918,11 @@ def use_gpu(self):
             if isinstance(value, chainer.Variable):
                 value.to_gpu()
 
+    def use_chainerx(self, device='native:1'):
```
Comment (niboshi) [inline tests/chainer_tests/test_function_node.py]: Why `native:1` rather than `native:0`? 
How about refactoring the test to use `inject_backend_tests`?

File: tests/chainer_tests/test_function_node.py [hunk 14]
```diff
@@ -960,14 +970,29 @@ def check_grad(self):
             self._print_variables('gxs (expected)', expected)
             raise
 
+    def chainerx_grad(self, device):
+        if self.loss_scale:
+            pytest.skip('Unsupported configurations')
```
Comment (niboshi) [inline tests/chainer_tests/test_function_node.py]: Let's be more specific.
```suggestion
            pytest.skip('chainerx.grad does not support loss_scale')
```

File: tests/chainer_tests/test_function_node.py [hunk 15]
```diff
@@ -992,11 +1017,26 @@ def check_double_grad(self):
     def test_double_grad_cpu(self):
         self.check_double_grad()
 
+    def double_chainerx_grad(self, device):
+        if self.loss_scale:
+            pytest.skip('Unsupported configurations')
```
Comment (niboshi) [inline tests/chainer_tests/test_function_node.py]: ditto.

File: chainerx_cc/chainerx/backward.cc (hunk 11 above)
Comment (emcastillo) [reply chainerx_cc/chainerx/backward.cc]: grad.has_value() evaluates to false in the case that the corresponding input value was an array that is not part of the graph.

I rewrote the loop to be in the same style of the above ones and iterate input_grads and inputs with the same index.

File: chainerx_cc/chainerx/backward.cc [hunk 16]
```diff
@@ -621,13 +639,27 @@ std::vector<absl::optional<Array>> Grad(
         }
     }
 
-    BackwardImpl{inputs, outputs, actual_backprop_id, double_backprop, std::move(array_node_grad_map)}.Run();
+    // Push initial output grads, Run assigns to 1 in other case
```
Comment (niboshi) [inline chainerx_cc/chainerx/backward.cc]: A nitpick: to more easily understand that it refers to a function.

```suggestion
    // Push initial output grads, Run() assigns to 1 in other case
```

File: chainerx_cc/chainerx/backward.cc [hunk 17]
```diff
@@ -621,13 +639,27 @@ std::vector<absl::optional<Array>> Grad(
...
+
+    BackwardImpl{inputs, outputs, actual_backprop_id, double_backprop, std::move(array_node_grad_map), retain_grad}.Run();
 
     // input_grads may contain unmodified array bodies (nullptr) for arrays that are not included in the graph.
     // Those grads are returned as nullopt.
-    for (absl::optional<Array>& grad : input_grads) {
-        if (grad.has_value() && internal::GetArrayBody(*grad) == nullptr) {
-            grad = absl::nullopt;
+    for (size_t i = 0; i < input_grads.size(); ++i) {
+        if (input_grads[i].has_value()) {
```
Comment (niboshi) [inline chainerx_cc/chainerx/backward.cc]: Let's factor out the repeated expression `input_grads[i]`.

File: tests/chainer_tests/test_function_node.py [hunk 18]
```diff
@@ -915,10 +913,10 @@ def setUp(self):
             self._init_ones(self._to_grad_names(self.y_names))
             self.gys = None
 
-    def use_gpu(self):
+    def use_device(self, device):
         for value in six.itervalues(self.__dict__):
             if isinstance(value, chainer.Variable):
-                value.to_gpu()
+                value.to_device(backend.get_device(device))
```
Comment (niboshi) [inline tests/chainer_tests/test_function_node.py]: `get_device` is redundant.

```suggestion
                value.to_device(device)
```

File: tests/chainer_tests/test_function_node.py [hunk 19]
```diff
@@ -989,18 +993,29 @@ def check_double_grad(self):
...
-
-    @attr.gpu
-    def test_double_grad_gpu(self):
-        self.use_gpu()
-        self.check_double_grad()
+    def test_double_grad(self, backend_config):
+        if self.loss_scale and backend_config.xp is chainerx:
+            pytest.skip('chainerx.grad does not support loss_scale')
+        self.use_device(backend_config.device)
+        with chainer.using_device(backend_config.device):
```
Comment (niboshi) [inline tests/chainer_tests/test_function_node.py]: ditto.

File: tests/chainer_tests/test_function_node.py [hunk 20]
```diff
@@ -960,13 +965,12 @@ def check_grad(self):
...
-    @attr.gpu
-    def test_grad_gpu(self):
-        self.use_gpu()
-        self.check_grad()
+    def test_grad(self, backend_config):
+        if self.loss_scale and backend_config.xp is chainerx:
+            pytest.skip('chainerx.grad does not support loss_scale')
+        self.use_device(backend_config.device)
+        with chainer.using_device(backend_config.device):
+            self.check_grad()
```
Comment (niboshi) [inline tests/chainer_tests/test_function_node.py]: `chainer.gad` should not require `using_device`.
Let's apply `using_device` to where it is necessary, not to the whole check function.

Comment (emcastillo) [reply tests/chainer_tests/test_function_node.py]: There was a bug in function_node that appeared when adding multigpu tests.

File: chainerx_cc/chainerx/backward.cc (hunk 11 above)
Comment (emcastillo) [reply chainerx_cc/chainerx/backward.cc]: Comment updated with more specific content

File: chainerx_cc/chainerx/backward.cc [hunk 21]
```diff
@@ -621,13 +639,29 @@ std::vector<absl::optional<Array>> Grad(
...
+
+    for (size_t i = 0; i < input_grads.size(); ++i) {
+        absl::optional<Array>& input_grad = input_grads[i];
+        if (input_grad.has_value()) {
+            // If the allocated input_grad was not used during calculation due to a not connected graph
+            // the body was never initialized and it remains as null
+            if (internal::GetArrayBody(*input_grad) == nullptr) {
+                input_grad = absl::nullopt;
+                // Explicitly assign the grad to the input array if specified
+                // Input grads are not set by default during Backward run
```
Comment (niboshi) [inline chainerx_cc/chainerx/backward.cc]: It seems this comment should be placed in the next clause?
How about moving the previous one to this clause, too?

File: tests/chainerx_tests/unit_tests/test_backward.py [hunk 22]
```diff
@@ -555,3 +562,96 @@ def fprop(x0, x1):
...
+
+    expected_gxs = (
+        chainerx.full(shape, 4, dtype),)
+
+    # If variables go out of scope
+    # then intermediate grads are wiped.
+    # We need to do all the computations here
+    a = xs[0]*2
+    b = a+xs[1]
+    c = a+b
```
Comment (niboshi) [inline tests/chainerx_tests/unit_tests/test_backward.py]: A nitpick but let's put spaces around the operators, for readability and consistency.

File: tests/chainerx_tests/unit_tests/test_backward.py [hunk 23]
```diff
@@ -555,3 +562,96 @@ def fprop(x0, x1):
...
+    xs = (
+        chainerx.full(shape, 3, dtype).require_grad(),
+        chainerx.full(shape, 5, dtype).require_grad(),)
+
+    expected_gxs = (
+        chainerx.full(shape, 4, dtype),)
+
+    # If variables go out of scope
+    # then intermediate grads are wiped.
+    # We need to do all the computations here
```
Comment (niboshi) [inline tests/chainerx_tests/unit_tests/test_backward.py]: Does this comment mean this test can't use `_check_grad`?
Please explicitly write that.
I looked at this test first, and couldn't understand what this comment means.

Comment (emcastillo) [reply tests/chainerx_tests/unit_tests/test_backward.py]: Exactly, I will rewrite the comment in a more understandable way.

**Comment by niboshi** [comment]
Jenkins, test this please

**Comment by niboshi** [comment]
Please fix the clang-tidy error:
```
20:38:25 Use -header-filter=.* to display errors from all non-system headers. Use -system-headers to display errors from system headers as well.
20:38:25 /repo/chainerx_cc/chainerx/backward.cc:632:5: error: use range-based for loop instead [modernize-loop-convert,-warnings-as-errors]
20:38:25     for (size_t i = 0; i < inputs.size(); ++i) {
20:38:25     ^   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
20:38:25         (auto input : inputs)
```
(but it's better to avoid `auto`).

**Comment by niboshi** [comment]
Jenkins, test this please

**Comment by emcastillo** [comment]
Fixed NOLINT issue of jenkins.

File: chainerx_cc/chainerx/backward.cc [hunk 24]
```diff
@@ -650,7 +650,7 @@ std::vector<absl::optional<Array>> Grad(
 
     BackwardImpl{inputs, outputs, actual_backprop_id, double_backprop, std::move(array_node_grad_map), retain_grad}.Run();
 
-    for (size_t i = 0; i < input_grads.size(); ++i) {  // NOLINT(modernize-loop-convert)
+    for (size_t i = 0; i < input_grads.size(); ++i) {
```
Comment (niboshi) [inline chainerx_cc/chainerx/backward.cc]: Hm, why did you remove `NOLINT`?

Comment (emcastillo) [reply chainerx_cc/chainerx/backward.cc]: I mistook the loop, it was the first one of the method, which I rewrote to use the for each notation.
I cross-checked this running clang-tidy in my environment.

Comment (niboshi) [reply chainerx_cc/chainerx/backward.cc]: Oh, I see!

**Comment by niboshi** [comment]
Jenkins, test this please
